### PR TITLE
feat: introduce provenance for derivation candidates

### PR DESCRIPTION
**Pull request by RaitoBezarius**
In the future, we will be able to bolt on more fields to take notes
which embedding has been used, which search term has been used, etc,
etc.

For now, we have a provenance flags we can use to hint the frontend what
has been used.

The dance for the M2M is required because it's not trivial to migrate
everything.

Progress towards:

- #189
- #274

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

### CODE REVIEW

**Comment by erictapen** [comment]
Sounds promising to have. I'd think of ways how to visualise this in https://github.com/Nix-Security-WG/nix-security-tracker/issues/203

**Comment by RaitoBezarius** [comment]
> Sounds promising to have. I'd think of ways how to visualise this in #203

Yeah, this is only progress towards that goal, there's more to come to help you with hints :)

**Review by erictapen — APPROVED** [review]
The content of ProvenanceFlags looks sensible to me, even though I'm not sure how I would use the hardware and kernel constraints.

**Comment by RaitoBezarius** [comment]
> The content of ProvenanceFlags looks sensible to me, even though I'm not sure how I would use the hardware and kernel constraints.

Showing "Linux" / "macOS" / etc, I'd say?

File: src/website/shared/listeners/automatic_linkage.py [hunk 1]
```diff
@@ -10,19 +11,36 @@
...
 
-def produce_linkage_candidates(container: Container) -> set[NixDerivation]:
+class ProvenanceFlags(IntFlag):
+    PACKAGE_NAME_MATCH = auto()
+    VERSION_CONSTRAINT_INRANGE = auto()
+    VERSION_CONSTRAINT_OUTOFRANGE = auto()
+    NO_SOURCE_VERSION_CONSTRAINT = auto()
+    # Whether the hardware constraint is matched for this derivation.
+    HARDWARE_CONSTRAINT_INRANGE = auto()
+    KERNEL_CONSTRAINT_INRANGE = auto()
```
Comment (fricklerhandwerk) [inline src/website/shared/listeners/automatic_linkage.py:21]: Why do we have that type twice, and the other is strict?

Comment (RaitoBezarius) [reply src/website/shared/listeners/automatic_linkage.py:21]: Fixed.

**Pull request merged**